Automattic-Tracks-iOS/Event Logging/EventLogging.swift

```diff
@@ -0,0 +1,175 @@
+import Foundation
+import CommonCrypto
+import CocoaLumberjack
+import Sodium
+
+public class EventLogging {
+
+    /// Add a Log File to the list of events that need to be uploaded
+    public func enqueueLogForUpload(log: LogFile) throws {
+        try uploadQueue.add(log)
+
+        /// Notify observers of a new log in the queue
+        delegate.didQueueLogForUpload(log)
+
+        /// Restart the automatic upload queue when log files are added
+        uploadNextLogFileIfNeeded()
+    }
+
+    /// Maintains a list of events that need to be uploaded
+    private let uploadQueue: EventLoggingUploadQueue
+
+    /// Coordinates one-at-a-time file log dequeuing and upload
+    private let lock = NSLock()
+    private let processingQueue = DispatchQueue(label: "event-logging-upload")
+
+    /// Uploads Events
+    private let uploadManager: EventLoggingUploadManager
+
+    /// Data Source
+    let dataSource: EventLoggingDataSource
+
+    /// Delegate
+    let delegate: EventLoggingDelegate
+
+    public init(dataSource: EventLoggingDataSource,
+         delegate: EventLoggingDelegate,
+         fileManager: FileManager = FileManager.default
+    ) {
+        self.dataSource = dataSource
+        self.delegate = delegate
+
+        self.uploadManager = EventLoggingUploadManager(dataSource: dataSource, delegate: delegate)
+
+        self.uploadQueue = EventLoggingUploadQueue(
+            storageDirectory: dataSource.logUploadQueueStorageURL,
+            fileManager: fileManager
+        )
+
+        /// Start taking items off the queue and uploading them if needed
+        uploadNextLogFileIfNeeded()
+    }
+
+    /// Schedule encryption and upload for the next log file in the queue
+    public func uploadNextLogFileIfNeeded() {
+        DispatchQueue.global(qos: .background).async(execute: runUploadLogs)
+    }
+
+    /// Current enqueued log files
+    public var queuedLogFiles: [LogFile] {
+        return uploadQueue.items
+    }
+
+    /// Support adding additional time between requests if they are failing – reset after an hour to match the server
+    private var exponentialBackoffTimer = ExponentialBackoffTimer(minimumDelay: 2, maximumDelay: 3600)
+
+    /// The date that uploads will automatically resume after being paused due to failure
+    public var uploadsPausedUntil: Date? {
```

Comment:
>As I understand it, delay will be set during ExponentialBackoffTimer.increment but won't be changed even if the time has moved on, so technically the nextDate time might have came and passed, right? 

Correct – my thinking was to separate the backoff timer from the operation – the timer is used to check "is it time to do this thing yet?" but has no influence on when it actually runs – another process must check the timer (which is deliberately (though perhaps this is a bad decision) a "dumb" object – it does nothing for itself and must be controlled and read externally).

>Would that mean the upload is still paused until a time?

This ends up being decided by the `EventLogging` class and the kernel – `EventLogging` asks the kernel to schedule the upload at the time specified by `next`, rather than the timer being responsible for invoking the task. 

>Again, I don't know how uploadsPausedUntil is going to be used, and I kind of wish we didn't need to expose this at all. However assuming it's necessary, maybe we can just compare the current date to the ExponentialBackoffTimer.nextDate which is a little more expensive but it is also much simpler and reliable.

It only exists to provide external debugging, and I'm good with that change ca1d5dc432154de47876541338775935fe67fa69.